python_bindings/src/builtin_helpers_src.py

```diff
@@ -0,0 +1,787 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from functools import total_ordering
+from inspect import isclass
+from typing import Any, Optional
+import sys
+
+# 'print()' is consumed by `halide.print()` which is implicitly present;
+# here's our quick-n-dirty wrapper for debugging:
+# def _print(*args):
+#     def write(data):
+#         sys.stdout.write(str(data))
+#     for i, arg in enumerate(args):
+#         if i:
+#             write(" ")
+#         write(arg)
+#     write('\n')
+
+# Everything below here is implicitly in the `halide` package
+
+
+def _fail(msg: str):
+    raise RuntimeError(msg)
```

Comment:
It might be polite to use a custom exception type like `GeneratorError` for the sake of metaprogramming systems built on top of this.

Reply:
Agreed, but I took a quick stab at this and it makes some of the tests hard to get right, since errors thrown by the pybind11 layer (e.g. "cannot cast foo to bad") are always RuntimeError. I'll make this a TODO for now.

Reply:
Update: I decided this was fairly doable after more examination -- ended up adding a base `hl.Error` class for the generic bindings, with `hl.GeneratorError` as a subclass of it used by the Generator-specific code.